### PR TITLE
Add option to list all resources when verifying service connector config

### PR DESCRIPTION
**Pull request by stefannica**
## Describe changes
Adds support to list all resources when verifying a service connector config. This is to make it possible for the dashboard to fetch all the information it needs to render the list of possible resources that can be accessed by a multi-type service connector (a service connector that can access multiple types of resources, like an AWS connector that can access S3 buckets, EKS clusters and ECR registries).

The core change of this PR is allowing resources from multiple types to be listed in the same `ServiceConnectorResourcesModel` object:

* previously:
```json
{
  "id": null,
  "name": null,
  "connector_type": {},
  "resource_type": "aws-generic",
  "resource_ids": [
        "us-east-1"
      ],
  "error": null,
   ...
}
```

* currently:
```json
{
  "id": null,
  "name": null,
  "connector_type": {},
  "resources": [
    {
      "resource_type": "aws-generic",
      "resource_ids": [
        "us-east-1"
      ],
      "error": null
    },
    {
      "resource_type": "s3-bucket",
      "resource_ids": [
        "s3://sagemaker-studio-907999144431-m11qlsdyqr8",
        "s3://sagemaker-studio-d8a14tvjsmb"
      ],
      "error": null
    },
   ...
}
```

This is an example of how the dashboard uses this new format returned in the `/verify` endpoint:

Register Connector Configuration - Stage 2

The same is reflected in the CLI now:

```
$ zenml service-connector verify aws-connectors
Service connector 'aws-connectors' is correctly configured with valid credentials and has access to the following resources:
┏━━━━━━━━━━━━━━━━━━━━━━━┯━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┓
┃     RESOURCE TYPE     │ RESOURCE NAMES                               ┃
┠───────────────────────┼──────────────────────────────────────────────┨
┃    🔶 aws-generic     │ us-east-1                                    ┃
┠───────────────────────┼──────────────────────────────────────────────┨
┃     📦 s3-bucket      │ s3://aws-ia-mwaa-715803424590                ┃
┃                       │ s3://zenbytes-bucket                         ┃
┃                       │ s3://zenfiles                                ┃
┃                       │ s3://zenml-demos                             ┃
┃                       │ s3://zenml-generative-chat                   ┃
┃                       │ s3://zenml-public-datasets                   ┃
┃                       │ s3://zenml-public-swagger-spec               ┃
┃                       │ s3://zenml-sandbox-infra                     ┃
┃                       │ s3://zenml-terraform-ci                      ┃
┠───────────────────────┼──────────────────────────────────────────────┨
┃ 🌀 kubernetes-cluster │ zenhacks-cluster                             ┃
┠───────────────────────┼──────────────────────────────────────────────┨
┃  🐳 docker-registry   │ 715803424590.dkr.ecr.us-east-1.amazonaws.com ┃
┗━━━━━━━━━━━━━━━━━━━━━━━┷━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┛
```

## Pre-requisites
Please ensure you have done the following:
- [x] I have read the **CONTRIBUTING.md** document.
- [ ] If my change requires a change to docs, I have updated the documentation accordingly.
- [ ] If I have added an integration, I have updated the [integrations](https://docs.zenml.io/component-gallery/integrations) table and the [corresponding website section](https://zenml.io/integrations).
- [ ] I have added tests to cover my changes.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Other (add details above)

### CODE REVIEW

File: src/zenml/client.py [hunk 1]
```diff
@@ -4538,8 +4541,9 @@ def create_service_connector(
                 "Python packages and ZenML integrations and try again."
             )
 
-        if not resource_type and len(connector.resource_types) == 1:
-            resource_type = connector.resource_types[0].resource_type
+        if not resource_type:
+            if len(connector.resource_types) == 1:
+                resource_type = connector.resource_types[0].resource_type
```
Comment (fa9r) [inline src/zenml/client.py:4546]: Why was this changed? Old syntax seemed cleaner IMO

File: src/zenml/client.py [hunk 2]
```diff
@@ -4768,8 +4784,9 @@ def update_service_connector(
         else:
             resource_types = resource_type
 
-        if not resource_type and len(connector.resource_types) == 1:
-            resource_types = connector.resource_types[0].resource_type
+        if not resource_type:
+            if len(connector.resource_types) == 1:
+                resource_types = connector.resource_types[0].resource_type
```
Comment (fa9r) [inline src/zenml/client.py:4789]: same here

File: src/zenml/models/service_connector_models.py [hunk 3]
```diff
@@ -857,6 +857,39 @@ def is_satisfied_by(
...
+        title="The type of resource that the service connector instance can "
+        "be used to access.",
+        max_length=STR_FIELD_MAX_LENGTH,
+    )
+
+    resource_ids: Optional[List[str]] = Field(
+        default=None,
+        title="The resource IDs of all resource instances that the service "
+        "connector instance can be used to access. Omitted (set to None) for "
+        "multi-type service connectors that didn't explicitly requested to "
```
Comment (fa9r) [inline src/zenml/models/service_connector_models.py:877]:
```suggestion
        "multi-type service connectors that didn't explicitly request to "
```

File: src/zenml/models/service_connector_models.py [hunk 4]
```diff
@@ -882,30 +915,92 @@ class ServiceConnectorResourcesModel(BaseModel):
...
-        "connector instance can be used to access. Omitted (set to None) for "
-        "multi-type service connectors. Also omitted if an error occurred "
-        "while listing the resource instances or if no resources are listed "
-        "due to authorization issues or lack of permissions (in both cases the "
-        "'error' field is set to an error message). For resource types that "
-        "do not support multiple instances, a single resource ID is listed.",
+    resources: List[ServiceConnectorTypedResourcesModel] = Field(
+        default_factory=list,
+        title="The list of resources that the service connector instance can "
+        "be used to access. Contains one entry for every resource type "
```
Comment (fa9r) [inline src/zenml/models/service_connector_models.py:921]:
```suggestion
        "give access to. Contains one entry for every resource type "
```

File: src/zenml/models/service_connector_models.py [hunk 5]
```diff
@@ -882,30 +915,92 @@ class ServiceConnectorResourcesModel(BaseModel):
...
+            resource.error = error
+            resource.resource_ids = None
+            if len(self.resources) == 1:
+                # If there is only one resource type involved, set the global
+                # error message as well.
+                self.error = error
+        else:
+            self.error = error
+            for resource in self.resources:
+                resource.error = error
```
[the remark this answers is not in the document]
Comment (fa9r) [reply src/zenml/models/service_connector_models.py:981]: If it does need to happen, maybe consider splitting the function into `set_global_error` and `set_resource_error` or similar

File: src/zenml/models/service_connector_models.py [hunk 6]
```diff
@@ -929,46 +1024,36 @@ def emojified_connector_type(self) -> str:
 
         return self.connector_type
 
-    @property
-    def emojified_resource_types(self) -> Optional[List[str]]:
-        """Get the emojified resource types.
+    def get_emojified_resource_types(
```
Comment (fa9r) [inline src/zenml/models/service_connector_models.py:1027]: **emojified**, love it ^^  🌺🌈🎉🌞🤗

File: src/zenml/service_connectors/service_connector.py [hunk 7]
```diff
@@ -1128,98 +1129,149 @@ def verify(
...
+            except Exception as exc:
+                error = (
+                    f"connector{name_msg} verification failed with "
+                    f"unexpected error: {exc}"
+                )
+                # Log an exception if debug logging is enabled
+                if logger.isEnabledFor(logging.DEBUG):
+                    logger.exception(error)
+                else:
+                    logger.warning(error)
```
Comment (fa9r) [inline src/zenml/service_connectors/service_connector.py:1229]: Seems odd, why?

**Review by fa9r — APPROVED** [review]
Looks good to me, UX-wise this is a great improvement for multi-type connectors, looking forward to getting this live 👍

**Comment by fa9r** [comment]
Oh, also, I'd love if we could add a unit test for the `list_resources` behavior, and maybe another one for `print_service_connector_resource_table` that checks that the result looks reasonable for both options of `show_resources_only`.

File: src/zenml/client.py (hunk 1 above)
Comment (stefannica) [reply src/zenml/client.py:4546]: probably the result of going through several stages to get this right. I'll change it back.

File: src/zenml/client.py (hunk 2 above)
Comment (stefannica) [reply src/zenml/client.py:4789]: I'll change it back

File: src/zenml/client.py [hunk 8]
```diff
@@ -4849,24 +4867,25 @@ def update_service_connector(
...
-
-        # For resource types that don't support multi-instances, it's
-        # better to save the default resource ID in the connector, if
-        # available. Otherwise, we'll need to instantiate the connector
-        # again to get the default resource ID.
-        if connector_resources:
+            # For resource types that don't support multi-instances, it's
+            # better to save the default resource ID in the connector, if
+            # available. Otherwise, we'll need to instantiate the connector
+            # again to get the default resource ID.
```
[the remark this answers is not in the document]
Comment (stefannica) [reply src/zenml/client.py:4880]: yes, you're correct, the "create service connector" and "update service connector" methods contain code that is very similar. However, the differences are subtle enough and spread out across several lines of code, which would make the code unreadable if I were to extract the common pieces out into a separate parameterized function.

File: src/zenml/service_connectors/service_connector.py (hunk 7 above)
Comment (stefannica) [reply src/zenml/service_connectors/service_connector.py:1229]: This is a choice I had to make to keep the CLI output cleaner. Basically, if you have a lot of connectors that spit out errors, you end up with screens full of exception traces and error messages before you get the final result, which can be really annoying. With this in place, you only get the final messages as warnings.

**Comment by stefannica** [comment]
> Oh, also, I'd love if we could add a unit test for the `list_resources` behavior, and maybe another one for `print_service_connector_resource_table` that checks that the result looks reasonable for both options of `show_resources_only`.

I'll see what I can do. It's not easy to implement unit-testing for any of this, given that we rely on real-life credentials and external services for the authentication part. I need to build something like an entirely fake service connector implementation. Definitely worth doing, but something that will take time and should be the subject of a dedicated PR.

**Review by schustmi — CHANGES_REQUESTED** [review]

File: src/zenml/service_connectors/service_connector.py [hunk 9]
```diff
@@ -1128,98 +1129,149 @@ def verify(
...
-                logger.warning(resources.error)
-            return resources
+            for rt in resource_types
+        ]
 
         if self.has_expired():
-            resources.error = (
-                "the connector's authentication credentials have expired."
-            )
+            error = "the connector's authentication credentials have expired."
```
[the remark this answers is not in the document]
Comment (stefannica) [reply src/zenml/service_connectors/service_connector.py:1154]: Yes, it's intended behavior: this flag only controls whether the `resource_ids` fields are populated or not, but the `resources` list is always included. I changed the method docstring to make that a bit more clear.

File: src/zenml/service_connectors/service_connector.py [hunk 10]
```diff
@@ -1128,98 +1129,149 @@ def verify(
...
-            resources.error = (
-                "an unexpected error occurred while verifying the connector."
-            )
-            return resources
+                resource_ids = self._verify(
+                    resource_type=resource_type,
+                    resource_id=resource_id,
+                )
+            except ValueError as exc:
+                logger.debug(f"connector verification failed: {exc}")
```
[the remark this answers is not in the document]
Comment (stefannica) [reply src/zenml/service_connectors/service_connector.py:1200]: you're right, we don't need it

**Review by schustmi — APPROVED** [review]
